Encryption.md

```diff
@@ -132,49 +203,164 @@ the encryption buffer is comprised of an IV, ciphertext and tag, as described in
 Algorithms section. The length of the encryption buffer (a 4-byte little endian) is 
 written to the output stream, followed by the buffer itself.
 
+|length (4 bytes) | IV (12 bytes) | ciphertext (`length` bytes) | tag (16 bytes) |
+|-----------------|---------------|-----------------------------|----------------|
+
 The column pages (data and dictionary) are encrypted after compression. For each page, 
-the encryption buffer is comprised of an IV, ciphertext and (in case of AES_GCM_V1) of a 
+the encryption buffer is comprised of an IV (nonce), ciphertext and (in case of AES_GCM_V1) of a 
 tag, as described in the Algorithms section. Only the buffer is written to the output 
-stream - not need to write the length of the buffer, since the length (size of the page after
+stream - no need to write the length of the buffer, since the length (page size after
 compression and encryption) is kept in the page headers.
 
-A `crypto_meta_data` field in set in each `ColumnChunk` in the encrypted columns. 
+| IV (12 bytes) | ciphertext (`page_size` bytes) | tag (16 bytes) |
+|---------------|--------------------------------|----------------|
+
+or 
+
+| IV (16 bytes) | ciphertext (`page_size` bytes) | 
+|---------------|--------------------------------|
+
+
+A `crypto_meta_data` field is set in each `ColumnChunk` in the encrypted columns. 
 `ColumnCryptoMetaData` is a union - the actual structure is chosen depending on whether the 
 column is encrypted with the footer key, or with a column-specific key. For the latter, 
 a key metadata can be specified, with a maximal length of 512. Key metadata is a free-form
-byte array that can be used by a reader to retrieve the column encryption key. 
-
-Parquet file footer, and its nested structures, contain sensitive information - ranging 
+byte array that can be used by a reader to retrieve the encryption key (see next subsection). 
+
+```c
+struct EncryptionWithFooterKey {
+}
+
+struct EncryptionWithColumnKey {
+  /** Column path in schema **/
+  1: required list<string> path_in_schema
+  
+  /** Retrieval metadata of the column-specific key **/
+  2: optional binary column_key_metadata
+}
+
+union ColumnCryptoMetaData {
+  1: EncryptionWithFooterKey ENCRYPTION_WITH_FOOTER_KEY
+  2: EncryptionWithColumnKey ENCRYPTION_WITH_COLUMN_KEY
+}
+
+struct ColumnChunk {
+...
+  /** Crypto metadata of encrypted columns **/
+  8: optional ColumnCryptoMetaData crypto_meta_data
+}
+```
+
+The `path_in_schema` field in  `EncryptionWithColumnKey` structure is needed because the 
+column name (path) is kept inside the encrypted `ColumnMetaData`. The column name is required 
+to build the AAD for decryption of the `ColumnMetaData` (see above). Also, the column name can 
+be used for explicit passing of a decryption key for this column, without calling a key retrieval callback.
```

Comment:
FileMetaData stores the schema, and the order here must match the order in the schema. I don't think the `path_in_schema` field is needed and it should be removed. In addition, I think the AAD should be based on the ordinal anyway.

Reply:
The reason is the integrity chain of trust. 

1. If footer is encrypted, and we trust all entities that have the footer key, then no problem.

2. If footer is plaintext, or we don'y fully trust an entity that got the footer key (say to read plaintext columns in the file) - the following attack is possible: 
- ColumnMetaData of ColA are swapped with ColumnMetaData of ColB
- Column order is changed in file footer / schema
- Since column ordinal is not explicitly recorded in ColumnMetaData, we can't verify it after decryption
- The result: reader of ColA gets data of ColB (say both are of int type)

However, path_in_schema is explicitly recorded in ColumnMetaData. Therefore, we can compare it to the AAD and detect the swapping.

I agree we don't have to keep path_in_schema in ColumnCryptoMetaData. Its a convenience, though, similar to keeping it in the ColumnMetaData (that also can use the ordinal to find it - but I guess prefers to have it readily available).

Reply:
I have a nagging feeling using the column number instead of explicit name is less safe, and can enable data swapping attacks. But let me think more about this. Today is a weekend here, on Sunday I'll try to work out the details of an attack - and possible defenses.

In the meantime - can you say why do you prefer the column ordinal to the name (path)? We consider an untrusted footer situation, where the footer fields (eg schema, order of column chunks, etc) can be modified by an attacker. ColumnMetaData can be swapped together with the column chunks - but the content integrity of ColumnMetaData is protected (by column key), and also, we assign an AAD to verify its "identity".

Reply:
The analysis has led to discovery of a more substantial threat. If the footer is untrusted - an attacker can remove full row groups (if there is more than one) - their pages, and/or their metadata inside the footer. Column encryption keys can't protect against this, because this is not column specific. The footer encryption key can.
Therefore, we should actually simplify the quality of service model. Plaintext footer mode doesn't guarantee data integrity protection. Users that care about data integrity attacks, must use the Encrypted footer mode.
Regarding column ordinal vs path - they offer the same level of integrity protection in the encrypted footer mode, since the schema can't be modified.

Reply:
The footer (or the whole file) could still be signed to prevent malicious modifications if we wanted to improve plain-text footer mode. The question is, how much extra complexity is this compatibility-mode worth? Personally I think that plain-text footer mode already complicates the design significantly while providing only questionable benefits. On the other hand, relative to the already existing plain-text footer design, adding a signature would only make it marginally more complicated.

Reply:
> instead of a sudden and painful migration from fully unprotected data - to fully protected data, the users can make the transition gradual, via an intermediate stage

Actually this is possible without a plain text footer mode as well, just takes more time:

- With the plain text footer mode, you first enable writing partially encrypted files, then update all your workflows to support encryption.
- Wihtout plain text footer mode, first you have to update all of your workflows to support encryption and only then you can enable encryption for writing new files.

Still, we have only reached an intermediate stage in both cases, because now some of the data are encrypted, some of it not. Phasing out unencrypted files is a separate step after this that would happen the same way in both cases.

Reply:
In general, yes, but the user's organization had not be willing to upgrade the workflow parts that don't touch the sensitive data.

Regarding the footer signature proposal - the more I think about it, the more I like it. The writer of encrypted files is connected to KMS. Hence it can get a "footer key" - but in a plaintext footer mode, will use it to sign the hash of the footer. The signed hash is written before the footer in the file, followed by the plaintext footer.  The legacy and new readers that parse plaintext columns only, don't care about this field. However, the new readers that parse sensitive columns - are connected to KMS, and have no problem getting the "footer key" and checking the footer integrity. 
Yes, it makes the model a bit more complex, but for a good reason. If this makes sense, lets add this feature.

Reply:
Or encrypt the serialized footer using AES GCM with the footer key and the AAD - and save the GCM tag (16 bytes) only. This is enough to verify footer integrity at the reader.

Reply:
I like the idea of using AES-GCM to produce an AAD and writing just the tag and not the encrypted footer. That's a great idea that doesn't require a lot of additional structures or specification.

Reply:
> can you say why do you prefer the column ordinal to the name (path)?

The security provided is identical, but you end up with an AAD suffix that is smaller to process and simpler to build.